### PR TITLE
Feature/docs improvements

### PR DESCRIPTION
**Pull request by dsyme**

### CODE REVIEW

File: src/fsharp/FSharp.Core/array.fsi [hunk 1]
```diff
@@ -8,6 +8,10 @@ namespace Microsoft.FSharp.Collections
     open System.Collections.Generic
 
     /// <summary>Basic operations on arrays.</summary>
+    ///
+    /// <remarks>
+    ///  See also <a href="https://docs.microsoft.com/dotnet/fsharp/language-reference/arrays">F# Language Guide - Arrays</a>.
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/array.fsi:13]: @cartermp The FSHarp.Core API docs are vastly improved if they can cross-reference back to https://docs.microsoft.com/dotnet/fsharp/language-reference topics, e.g. for sequences, async, lists, arrays and so on.

What's the right URL we should use?  Just direct linking to the doc section seems to give the best user experience.   Setting up aka.ms links seems kind of wrong for an open source component but if you think we should do it we can.

**Comment by dsyme** [comment]
> @KevinRansom deleted the feature/docs branch 9 hours ago


@KevinRansom As mentioned here please don't delete feature/docs branch (or tell me a better name to use please?)

https://github.com/dotnet/fsharp/pull/9844#issuecomment-667445274

**Comment by cartermp** [comment]
This was @KevinRansom, it was a setting enabled so that stale branches just get removed when a PR gets merged.

Is it necessary to follow this pattern with a separate branch? Why can't `master` just be used?

**Comment by cartermp** [comment]
FWIW I _really_ want to cut down on the noise generated by all these branches with CI. We set it up specifically for releases just so that we don't mess up parallel releases into VS, not as a generalized mechanism to enable CI and bot traffic.

**Comment by dsyme** [comment]
> Is it necessary to follow this pattern with a separate branch? Why can't master just be used?

Well, we could.  But do you really want to?

In my experience the process of building, refining and releasing docs always results in last minute improvements to the docs, especially when writing and refining hundreds of examples.   Literally hundreds of commits may flow into that branch as people refine the documentation and proof-read it.  

This is of course an integral problem of having the documentation in the source code itself.  

So I guess I'm forethinking the likelihood that we'd start to get feedback of "why are you flooding master with so many doc requests".  And perhaps an expectation that people will be much more motivated to work on docs if they can make it a separate activity, not flooding master branch or carrying the risk of damage through a typo, and then integrated across in appropriate batches.

Additionally, if people contribute to feature/docs we would almost certainly approve immediately based on a visual scan.  We actually don't even need to run CI except when doing feature/fsdocs --> master integrations.

Note this is also a reason why the compiler guide is not hosted in this repository - it could but that may result in a lot of flooding of master as edits are made.


> FWIW I really want to cut down on the noise generated by all these branches with CI. 

I understand this motivation. It's a shame there's so much noise, because the mechanism of auto-integrations is incredibly useful otherwise.

**Comment by cartermp** [comment]
> Literally hundreds of commits (including from non-engineers) may flow into the docs branch as people refine the documentation and proof-read it.

This sounds like the kind of thing we definitely _don't_ want to have to deal with in various integrations into the main branch. We try very hard to ensure all work goes into the main branch and flows into others as needed, but this would imply either a reverse-integration step that would surely lead to merge conflicts that can complicate things at a critical time, or bidirectional bot PRs that flow changes back and forth. Neither sounds good to me.

Given that the FSharp.Core XML docs must ultimately contain improvements, I don't see why we wouldn't accept many PRs that are trivial to merge into master and have everything benefit. In fact, we've gotten quite good (generally) at handling small PRs that improve things in an obvious manner and are low impact, and we love these kinds of contributions because they improve the world and are not risky to integrate at any stage of our release cycle.

**Comment by jonsequitur** [comment]
To the extent that automated documentation verification, generation, and publication can be improved over time, keeping the docs in the master branch has a big potential maintainability upside. For example, `dotnet-try` can be used right now to ensure that code sample actually compile, and to capture the actual code output into the documetnation.

If PR volume is an issue, we could presort them using labels and perhaps a bot to avoid the branch maintenance.

**Comment by cartermp** [comment]
Completely agree @jonsequitur - we should have these in the main branch by default unless there is a very, very good reason to have a separate feature branch. TBH I don't see one. Even the stated rationale of "making it easy for contributors" feels flimsy. Expected workflows in repos are that you target the main branch when doing work, not a special branch for special kinds of contributions.

**Comment by dsyme** [comment]
> To the extent that automated documentation verification, generation, and publication can be improved over time, keeping the docs in the master branch has a big potential maintainability upside

Hmmm I think there's some confusion here. 

As you can see from this and #9880 I'm just using `feature/docs` as a stable place to push doc improvements to master from.  I could just as well do this from any branch, it's just useful to have a stable name.  However the doc improvements are heading into master (of course), and on the steady state feature/docs is identical to master.  There is an automatic master --> feature/docs integration set up to maintain one side of that.

But honestly, it's really no big deal.  It's just more efficient this way (without any divergence).

In the set up I'd prefer, feature/docs is allowed to be a few commits ahead of master (consisting only of doc improvements) - and there should be an active pull request (like this one, or #9880) to push them into master. This allows the doc writer or the person building the published docs for FSharp.Core (e.g. at https://github.com/fsharp/fsharp-core-docs/) to push corrections and improvements (of the kind which inevitably get spotted while preparing docs), but not have publishgin blocked on those being merged to master.  It gives a stable point to publish docs from that is actively allowed to be a few "doc correction and improvement" commits ahead of master.

It is actually fairly normal that a doc branch is *ahead* of the branch being documented (and continuously integrated with it). That's how we always do docs at Microsoft isn't it?  I see doc merges from doc teams all the time.  It's exactly that set up.

**Comment by cartermp** [comment]
> That's how we always do docs at Microsoft isn't it?

Not to my knowledge, no.

Sorry for the misunderstanding here. As you can tell, I'm _very_ allergic to any kind of long-term "docs" branch that could possibly diverge.

A stable branch name is fine, but what we shouldn't do is increase bot traffic by having master merge into it. These things are dreadful noise that we only keep for essential purposes and ultimately want to be rid of for most things:

image

Normal means of contributing code, spelling fixes, formatting improvements, etc. are just by branching off of master and submitting a PR that does this. We shouldn't specialize improvements to XML documentation for FSharp.Core. PRs that aren't long-poll, big features tend to get merged pretty quickly in this repo. Is the concern rooted in that not being the case?

**Comment by dsyme** [comment]
> Is the concern rooted in that not being the case?

To be honest for the moment I just want a stable branch that 

1. Is always there (so I and fsharp-core-docs CI can always build the latest and best FSharp.Core docs from it)
2. Is automatically kept up-to-date with master through master --> branch merges (I actually don't care if CI is run on that)
3. I can freely push doc improvements to while improving the docs, not gated
4. We manually integrate back into master regularly (a normal PR)

I do wish we could just get rid of the noise.  Maybe there's a way to turn of CI for the master --> branch merges in the case of this specific branch.

**Pull request merged**